README.md
emu2149
=======

A YM2149 (aka PSG) emulator written in C.

# Breaking Change on v1.40
- Internal clock divider is disabled by default. Use `PSG_setClockDivider(psg, 1)` to enable clock divider.

```
/* Typical YM2149 configuration */
PSG *psg = PSG_new(3579545, 44100);
PSG_setClockDivider(psg, 1);
PSG_setVolumeMode(psg, YM2149_PERFECT); // YM style
PSG_reset(psg);

/* Typical AY-3-8910 configuration */
PSG *psg = PSG_new(3579545/2, 44100);
PSG_setVolumeMode(psg, AY8910_ZXTUNE); // AY style
PSG_reset(psg);
```

# Note for rate conversion

The internal sample rate converter in emu2149 is a light-weight, simplified procedure.
Even if `1` is set to the quality (PSG_setQuality), the alias noise cannot be completely eliminated.

To obtain the best accurate result, an external sample rate conversion should be used. 
Please follow these steps.

1. Set the sample rate to 1/8 of the chip's master clock. i.e. rate=223721Hz for master=1.78MHz.
2. Set `0` to the quality value with `PSG_setQuality(psg, 0)` so that you can bypass the internal rate converter.
3. Call `PSG_calc()` to generate samples.
4. Apply your preferred external rate converter to the samples to obtain your desired sample rate.

CHANGELOG.md
## 2023-03-19 :
- Now noise emulation is from MAME, confirmed to be completely compatible with real YM2149 and YM6630 hardware.
- Volume tables are now available in several different flavors - from ZXTune, AYumi or perfectly logarithmic (the original was abolished due to being the worst ever), so envelope emulation is now also accurate (and not with peaks 2x thicker than they should be), also they are now doubles instead of bytes.

## 2022-11-26 : v1.41
- Fix a problem with DC offset when muting a channel (issue #4).
- Disable frequency limiter if the internal rate converter is not active (i.e. quality == 0).

## 2022-11-25 : v1.40
Breaking Change: ym2149's internal clock divider is disabled by default.
Use PSG_setClockDivider to enable clock divider.

- Fix the bug with the changing period register faster than frequency (issue #3).
- Fix the confusing PSG_setVolumeMode type argument.
- Rename PSG_set_rate to PSG_setRate.
- Rename PSG_set_quality to PSG_setQuality.
- Restore PSG_setClock that was removed on v1.20 for compatibility.
- Add PSG_setClockDivider function.
- Fix noise period at zero frequency.

## 2021-09-29 : v1.30
- Fix some envelope generator problems (issue #2).

## 2016-09-06 : v1.20
- Support per-channel output.

## 2015-12-13 : v1.17
- Changed own integer types to C99 stdint.h types.

## 2004-01-11 : v1.16
- Fixed the envelope problem where the envelope frequency register is written before key-on.

## 2003-09-19 : v1.15
- Added PSG_setMask and PSG_toggleMask.

## 2002-10-13 : v1.14
- Fixed the envelope unit.

## 2002-03-02 : v1.12
- Removed PSG_init & PSG_close.

## 2001-10-03 : v1.11     
- Added PSG_set_quality().

## 2001-08-14 : v1.10
- Bump Version.

## 2001-04-28 : v1.00 beta 
- 1st Beta Release.










emu2149.h
/* emu2149.h */
#ifndef _EMU2149_H_
#define _EMU2149_H_

#include <stdint.h>

#define PSG_MASK_CH(x) (1<<(x))

#ifdef __cplusplus
extern "C"
{
#endif

  typedef struct __PSG
  {

    // /* temp shite */
    // char debug [32];
    /* Volume Tables */
    const double *voltbl;

    uint8_t reg[0x20];
    double out;

    uint32_t clk, rate, base_incr;
    uint8_t quality;
    uint8_t clk_div;

    uint16_t count[3];
    uint8_t volume[3];
    uint16_t freq[3];
    uint8_t edge[3];
    uint8_t tmask[3];
    uint8_t nmask[3];
    uint32_t mask;

    uint32_t base_count;

    uint8_t env_ptr;
    uint8_t env_face;

    uint8_t env_continue;
    uint8_t env_attack;
    uint8_t env_alternate;
    uint8_t env_hold;
    uint8_t env_pause;

    uint16_t env_freq;
    uint32_t env_count;

    uint32_t noise_lfsr;
    uint8_t noise_scaler;
    uint8_t noise_count;
    uint8_t noise_freq;

    /* rate converter */
    uint32_t realstep;
    uint32_t psgtime;
    uint32_t psgstep;

    uint32_t freq_limit;

    /* I/O Ctrl */
    uint8_t adr;

    /* output of channels */
    double ch_out[3];

  } PSG;

  void PSG_setQuality (PSG * psg, uint8_t q);
  void PSG_setClock(PSG *psg, uint32_t clk);
  void PSG_setClockDivider(PSG *psg, uint8_t enable);
  void PSG_setRate (PSG * psg, uint32_t rate);
  PSG *PSG_new (uint32_t clk, uint32_t rate);
  void PSG_reset (PSG *);
  void PSG_delete (PSG *);
  void PSG_writeReg (PSG *, uint32_t reg, uint32_t val);
  void PSG_writeIO (PSG * psg, uint32_t adr, uint32_t val);
  uint8_t PSG_readReg (PSG * psg, uint32_t reg);
  uint8_t PSG_readIO (PSG * psg);
  double PSG_calc (PSG *);
  void PSG_setVolumeMode (PSG * psg, int type);
  uint32_t PSG_setMask (PSG *, uint32_t mask);
  uint32_t PSG_toggleMask (PSG *, uint32_t mask);
    
#ifdef __cplusplus
}
#endif

/* deprecated interfaces */
#define PSG_set_quality PSG_setQuality
#define PSG_set_rate PSG_setRate
#define PSG_set_clock PSG_setClock

#endif


volumetables.h
#include "emu2149.h"

static const double voltbl[7][32] = {
  //Ayumi YM2149
  {
  0.0, 0.0,
  0.00465400167849, 0.00772106507973,
  0.0109559777218, 0.0139620050355,
  0.0169985503929, 0.0200198367285,
  0.024368657969, 0.029694056611,
  0.0350652323186, 0.0403906309606,
  0.0485389486534, 0.0583352407111,
  0.0680552376593, 0.0777752346075,
  0.0925154497597, 0.111085679408,
  0.129747463188, 0.148485542077,
  0.17666895552, 0.211551079576,
  0.246387426566, 0.281101701381,
  0.333730067903, 0.400427252613,
  0.467383840696, 0.53443198291,
  0.635172045472, 0.75800717174,
  0.879926756695, 1.0
},
  //ZXTune YM2149
  {
    0.0, 0.0,
    0.003646906233310, 0.007080186160067,
    0.010009918364233, 0.012359807736324,
    0.015350576028077, 0.018829633020523,
    0.023697261005570, 0.029327840085451,
    0.035431448844129, 0.039887083237964,
    0.047623407339590, 0.057465476462959,
    0.067139696345464, 0.076661326009003,
    0.091081101701380, 0.109208819714656,
    0.128450446326390, 0.146791790646220,
    0.174334325169757, 0.208972304875257,
    0.243244068055238, 0.278782330052644,
    0.332066834515908, 0.398931868467231,
    0.465583276112001, 0.532295719844358,
    0.632120241092546, 0.753948271915770,
    0.877485313191424, 1.0
},

  // //MAME YM2149
  // {
  //   0.0024286405418481, 0.0032123850991776,
  //   0.0042380261248186, 0.0055926463473633,
  //   0.0073730043541364, 0.0089501693275278,
  //   0.0108659893565554, 0.0135171746492501,
  //   0.0168069666182874, 0.0197677793904209,
  //   0.0232510885341074, 0.0281470730527334,
  //   0.0340686985970005, 0.0396226415094340,
  //   0.0460861151427189, 0.0550459603289792,
  //   0.0657571359458152, 0.0753459119496855,
  //   0.0863280116110305, 0.1016545718432511,
  //   0.1197000483792937, 0.1355491049830672,
  //   0.1534978229317852, 0.1784905660377359,
  //   0.2075568456700532, 0.2348234155781325,
  //   0.2656797290759555, 0.3108369617803580,
  //   0.3636768263183358, 0.5095016932752782,
  //   0.7137880986937590, 1.0000000000000000
  // },
  //Perfect logarithmical YM2149
  {
    0.0, 0.00552427172801990204,
    0.00656950324416964541, 0.0078125,
    0.00929068058595875802, 0.01104854345603980409,
    0.01313900648833929083, 0.015625,
    0.01858136117191751605, 0.02209708691207960818,
    0.02627801297667858166, 0.03125,
    0.03716272234383503209, 0.04419417382415921636,
    0.05255602595335716332, 0.0625,
    0.07432544468767006418, 0.08838834764831843271,
    0.10511205190671432663, 0.125,
    0.14865088937534012836, 0.17677669529663686543,
    0.21022410381342865326, 0.25,
    0.29730177875068025672, 0.35355339059327373086,
    0.42044820762685730653, 0.5,
    0.59460355750136051345, 0.70710678118654746172,
    0.84089641525371461306, 1.0
  },
  //Ayumi AY-3-8910
  {
    0.0, 0.0,
    0.00999465934234, 0.00999465934234,
    0.0144502937362, 0.0144502937362,
    0.0210574502174, 0.0210574502174,
    0.0307011520562, 0.0307011520562,
    0.0455481803616, 0.0455481803616,
    0.0644998855573, 0.0644998855573,
    0.107362478065, 0.107362478065,
    0.126588845655, 0.126588845655,
    0.20498970016, 0.20498970016,
    0.292210269322, 0.292210269322,
    0.372838941024, 0.372838941024,
    0.492530708782, 0.492530708782,
    0.635324635691, 0.635324635691,
    0.805584802014, 0.805584802014,
    1.0, 1.0
  },

  //ZXTune AY-3-8910
  {
    0.0, 0.0,
    0.0126955062181, 0.0126955062181,
    0.0185549706264, 0.0185549706264,
    0.0271305409323, 0.0271305409323,
    0.0401007095445, 0.0401007095445,
    0.0588845654994, 0.0588845654994,
    0.0822766460670, 0.0822766460670,
    0.1347371633478, 0.1347371633478,
    0.1586785687037, 0.1586785687037,
    0.2551003280690, 0.2551003280690,
    0.3559777218280, 0.3559777218280,
    0.4466620889600, 0.4466620889600,
    0.5639429312581, 0.5639429312581,
    0.7082627603571, 0.7082627603571,
    0.8422980086976, 0.8422980086976,
    1.0, 1.0
  },

  // //MAME AY-3-8910
  // {
  //   0.1470219435736677, 0.1470219435736677,
  //   0.1891536050156740, 0.1891536050156740,
  //   0.2473354231974922, 0.2473354231974922,
  //   0.3021943573667711, 0.3021943573667711,
  //   0.3752351097178683, 0.3752351097178683,
  //   0.4507836990595611, 0.4507836990595611,
  //   0.5385579937304075, 0.5385579937304075,
  //   0.6645768025078370, 0.6645768025078370,
  //   0.7128526645768025, 0.7128526645768025,
  //   0.8081504702194358, 0.8081504702194358,
  //   0.8539184952978056, 0.8539184952978056,
  //   0.8949843260188087, 0.8949843260188087,
  //   0.9253918495297806, 0.9253918495297806,
  //   0.9460815047021943, 0.9460815047021943,
  //   0.9623824451410659, 0.9623824451410659,
  //   1.0000000000000000, 1.0000000000000000
  // }

};

emu2149.c
/*
 * emu2149 v1.41
 * https://github.com/digital-sound-antiques/emu2149
 * Copyright (C) 2001-2022 Mitsutaka Okazaki
 *
 * Improvements done in 2023 by alexmush
 * This source refers to the following documents. The author would like to thank all the authors who have
 * contributed to the writing of them.
 * - psg.vhd        -- 2000 written by Kazuhiro Tsujikawa.
 * - s_fme7.c       -- 1999,2000 written by Mamiya (NEZplug).
 * - ay8910.c       -- 1998-2001 Author unknown (MAME).
 * - MSX-Datapack   -- 1991 ASCII Corp.
 * - AY-3-8910 data sheet
 * 
 * alexmush also referred to:
 * - YM2149 data sheet
 * - 2 real YM2149 chips
 */
#include <stdlib.h>
#include <string.h>
#include "emu2149.h"
#include "volumetables.h"


enum ChipTypes {
  YM2149_AYUMI = 0,
  YM2149_ZXTUNE = 1,
  // YM2149_MAME = 2,
  YM2149_PERFECT = 2,
  AY8910_AYUMI = 3,
  AY8910_ZXTUNE = 4,
  // AY8910_MAME = 6,
  VOLUME_TABLE_LAST = 4
};

static const uint8_t regmsk[16] = {
    0xff, 0x0f, 0xff, 0x0f, 0xff, 0x0f, 0x1f, 0x3f, 
    0x1f, 0x1f, 0x1f, 0xff, 0xff, 0x0f, 0xff, 0xff
};

#define GETA_BITS 24

static void
internal_refresh (PSG * psg)
{
  uint32_t f_master = psg->clk;
  
  if (psg->clk_div)
    f_master /= 2;

  if (psg->quality)
  {
    psg->base_incr = 1 << GETA_BITS;
    psg->realstep = f_master;
    psg->psgstep = psg->rate * 8;
    psg->psgtime = 0;
    psg->freq_limit = (uint32_t)(f_master / 16 / (psg->rate / 2));
  }
  else
  {
    psg->base_incr = (uint32_t)((double)f_master * (1 << GETA_BITS) / 8 / psg->rate);
    psg->freq_limit = 0;
  }
}

void 
PSG_setClock(PSG *psg, uint32_t clock)
{
  if (psg->clk != clock) {
    psg->clk = clock;
    internal_refresh(psg);
  }
}

void 
PSG_setClockDivider(PSG *psg, uint8_t enable)
{
  if (psg->clk_div != enable) {
    psg->clk_div = enable;  
    internal_refresh (psg);
  }
}

void
PSG_setRate (PSG * psg, uint32_t rate)
{
  uint32_t r = rate ? rate : 44100;
  if (psg->rate != r) {
    psg->rate = r;
    internal_refresh(psg);
  }
}

void
PSG_setQuality (PSG * psg, uint8_t q)
{
  if (psg->quality != q) {
    psg->quality = q;
    internal_refresh(psg);
  }
}

PSG *
PSG_new (uint32_t clock, uint32_t rate)
{
  PSG *psg;

  psg = (PSG *) calloc (1, sizeof (PSG));
  if (psg == NULL)
    return NULL;

  PSG_setVolumeMode(psg, 0);
  psg->clk = clock;
  psg->clk_div = 0;
  psg->rate = rate ? rate : 44100;
  psg->quality = 0;
  internal_refresh(psg);
  PSG_setMask(psg, 0x00);
  return psg;
}

void
PSG_setVolumeMode (PSG * psg, int type)
{
  if (type > VOLUME_TABLE_LAST)
    psg->voltbl = voltbl[0];
  else
    psg->voltbl = voltbl[type];
}

uint32_t
PSG_setMask (PSG *psg, uint32_t mask)
{
  uint32_t ret = 0;
  if(psg)
  {
    ret = psg->mask;
    psg->mask = mask;
  }  
  return ret;
}

uint32_t
PSG_toggleMask (PSG *psg, uint32_t mask)
{
  uint32_t ret = 0;
  if(psg)
  {
    ret = psg->mask;
    psg->mask ^= mask;
  }
  return ret;
}

void
PSG_reset (PSG * psg)
{
  int i;

  psg->base_count = 0;

  for (i = 0; i < 3; i++)
  {
    psg->count[i] = 0;
    psg->freq[i] = 0;
    psg->edge[i] = 0;
    psg->volume[i] = 0;
    psg->ch_out[i] = 0;
  }

  psg->mask = 0;

  for (i = 0; i < 16; i++)
    psg->reg[i] = 0;
  psg->adr = 0;


  psg->noise_lfsr = 1;
  psg->noise_scaler = 0;
  psg->noise_count = 0;
  psg->noise_freq = 0;

  psg->env_ptr = 0;
  psg->env_freq = 0;
  psg->env_count = 0;
  psg->env_pause = 1;

  psg->out = 0;

}

void
PSG_delete (PSG * psg)
{
  free (psg);
}

uint8_t
PSG_readIO (PSG * psg)
{
  return (uint8_t) (psg->reg[psg->adr]);
}

uint8_t
PSG_readReg (PSG * psg, uint32_t reg)
{
  return (uint8_t) (psg->reg[reg & 0x1f]);
}

void
PSG_writeIO (PSG * psg, uint32_t adr, uint32_t val)
{
  if (adr & 1)
    PSG_writeReg (psg, psg->adr, val);
  else
    psg->adr = val & 0x1f;
}

static inline void
update_output (PSG * psg)
{

  int i, noise;
  uint8_t incr;

  psg->base_count += psg->base_incr;
  incr = (psg->base_count >> GETA_BITS);
  psg->base_count &= (1 << GETA_BITS) - 1;

  /* Envelope */
  psg->env_count += incr;

  if (psg->env_count >= psg->env_freq)
  {
    if (!psg->env_pause)
    {
      if(psg->env_face)
        psg->env_ptr = (psg->env_ptr + 1) & 0x3f ; 
      else
        psg->env_ptr = (psg->env_ptr + 0x3f) & 0x3f;
    }

    if (psg->env_ptr & 0x20) /* if carry or borrow */
    {
      if (psg->env_continue)
      {
        if (psg->env_alternate^psg->env_hold) psg->env_face ^= 1;
        if (psg->env_hold) psg->env_pause = 1;
        psg->env_ptr = psg->env_face ? 0 : 0x1f;       
      }
      else
      {
        psg->env_pause = 1;
        psg->env_ptr = 0;
      }
    }

    if (psg->env_freq >= incr) 
      psg->env_count -= psg->env_freq;
    else
      psg->env_count = 0;
  }
  /* Noise */
  psg->noise_count += incr;
  if (psg->noise_count >= psg->noise_freq)
  {
    psg->noise_scaler ^= 1;
    if (psg->noise_scaler) 
    { 
      psg->noise_lfsr = ((psg->noise_lfsr & 0x1FFFE) >> 1) | (((psg->noise_lfsr&1) ^ ((psg->noise_lfsr>>3)&1)) << 16);
    }

    if (psg->noise_freq >= incr)
      psg->noise_count -= psg->noise_freq;
    else
      psg->noise_count = 0;
  }
  noise = ~(psg->noise_lfsr) & 1;
  /* Tone */
  for (i = 0; i < 3; i++)
  {
    psg->count[i] += incr;
    if (psg->count[i] >= psg->freq[i])
    {
      psg->edge[i] = !psg->edge[i];

      if (psg->freq[i] >= incr) 
        psg->count[i] -= psg->freq[i];
      else
        psg->count[i] = 0;
    }

    // if (0 < psg->freq_limit && psg->freq[i] <= psg->freq_limit) 
    // {
    //   /* Mute the channel if the pitch is higher than the Nyquist frequency at the current sample rate, 
    //    * to prevent aliased or broken tones from being generated. Of course, this logic doesn't exist 
    //    * on the actual chip, but practically all tones higher than the Nyquist frequency are usually 
    //    * removed by a low-pass circuit somewhere, so we here halt the output. */
    //   continue;
    // }

    if (psg->mask & PSG_MASK_CH(i)) 
    {
      psg->ch_out[i] = 0;
      continue;
    }

    if ((psg->tmask[i]||psg->edge[i]) && (psg->nmask[i]||noise))
    {
      if (!(psg->volume[i] & 32))
        psg->ch_out[i] = (psg->voltbl[(psg->volume[i] & 31) != 0 ? (psg->volume[i] & 31) + 1 : 0] / 4);
      else 
        psg->ch_out[i] = (psg->voltbl[psg->env_ptr] / 4);
    }
    else 
    {
      psg->ch_out[i] = 0;
    }
  }
}

static double 
mix_output(PSG *psg) 
{
  return (double)(psg->ch_out[0] + psg->ch_out[1] + psg->ch_out[2]);
}

double
PSG_calc (PSG * psg)
{
  if (!psg->quality) 
  {
    update_output(psg);
    psg->out = mix_output(psg);
  }
  else
  {
    /* Simple rate converter (See README for detail). */
    while (psg->realstep > psg->psgtime)
    {
      psg->psgtime += psg->psgstep;
      update_output(psg);
      psg->out += mix_output(psg);
      psg->out >>= 1;
    }
    psg->psgtime -= psg->realstep;
  }
  return psg->out;
}

void
PSG_writeReg (PSG * psg, uint32_t reg, uint32_t val)
{
  int c;

  if (reg > 15) return;

  val &= regmsk[reg];

  psg->reg[reg] = (uint8_t) val;

  switch (reg)
  {
  case 0:
  case 2:
  case 4:
  case 1:
  case 3:
  case 5:
    c = reg >> 1;
    psg->freq[c] = ((psg->reg[c * 2 + 1] & 15) << 8) + psg->reg[c * 2];
    break;

  case 6:
    psg->noise_freq = val & 31;
    break;

  case 7:
    psg->tmask[0] = (val & 1);
    psg->tmask[1] = (val & 2);
    psg->tmask[2] = (val & 4);
    psg->nmask[0] = (val & 8);
    psg->nmask[1] = (val & 16);
    psg->nmask[2] = (val & 32);
    break;

  case 8:
  case 9:
  case 10:
    psg->volume[reg - 8] = val << 1;
    break;
 
  case 11:
  case 12:
    psg->env_freq = (psg->reg[12] << 8) + psg->reg[11];
    break;

  case 13:
    psg->env_continue = (val >> 3) & 1;
    psg->env_attack = (val >> 2) & 1;
    psg->env_alternate = (val >> 1) & 1;
    psg->env_hold = val & 1;
    psg->env_face = psg->env_attack;
    psg->env_pause = 0;
    psg->env_ptr = psg->env_face ? 0 : 0x1f;
    break;

  case 14:
  case 15:
  default:
    break;
  }

  return;
}